### ISSUE TITLE
Solaar only shows 20%, 50%, and 90% levels for MX Ergo - show next level

### ISSUE THREAD

**Issue by khaytsus** [pwr-Solaar/Solaar#683]
**Information**

- Solaar version: 1.0.1
- Distribution: Fedora
- Kernel version (ex. `uname -srmo`): Linux 5.4.18-100.fc30.x86_64 x86_64 GNU/Linux
- Output of `solaar show`:

This is immediately after charging has completed.

```
Unifying Receiver
  Device path  : /dev/hidraw5
  USB id       : 046d:c52b
  Serial       : A027983A
    Firmware   : 12.09.B0030
    Bootloader : 04.16
    Other      : AA.AA
  Has 1 paired device(s) out of a maximum of 6.
  Notifications: wireless, software present (0x000900)
  Device activity counters: 1=112

  1: MX Ergo Multi-Device Trackball 
     Codename     : MX Ergo
     Kind         : trackball
     Wireless PID : 406F
     Protocol     : HID++ 4.5
     Polling rate : 8 ms (125Hz)
     Serial number: BD87365D
        Bootloader: BOT 49.00.B0001
          Firmware: MPM 06.03.B0022
             Other: 
     The power switch is located on the base.
     Supports 36 HID++ 2.0 features:
         0: ROOT                   {0000}   
         1: FEATURE SET            {0001}   
         2: DEVICE FW VERSION      {0003}   
         3: DEVICE NAME            {0005}   
         4: WIRELESS DEVICE STATUS {1D4B}   
         5: unknown:0007           {0007}   
         6: RESET                  {0020}   
         7: unknown:0021           {0021}   
         8: BATTERY STATUS         {1000}   
         9: LED CONTROL            {1300}   
        10: REPROG CONTROLS V4     {1B04}   
        11: unknown:1C00           {1C00}   
        12: unknown:2006           {2006}   
        13: POINTER SPEED          {2205}   
            Pointer Speed: 1.0
        14: VERTICAL SCROLLING     {2100}   
            Roller type: standard
            Ratchet per turn: 18
            Scroll lines: 0
        15: unknown:00C2           {00C2}   
        16: unknown:1802           {1802}   internal, hidden
        17: unknown:1803           {1803}   internal, hidden
        18: unknown:1806           {1806}   internal, hidden
        19: unknown:1805           {1805}   internal, hidden
        20: unknown:1813           {1813}   internal, hidden
        21: CHANGE HOST            {1814}   
        22: unknown:1815           {1815}   
        23: unknown:1830           {1830}   internal, hidden
        24: unknown:1861           {1861}   internal, hidden
        25: unknown:1890           {1890}   internal, hidden
        26: unknown:1891           {1891}   internal, hidden
        27: unknown:18A1           {18A1}   internal, hidden
        28: unknown:1DF3           {1DF3}   internal, hidden
        29: unknown:1E00           {1E00}   hidden
        30: unknown:1EB0           {1EB0}   internal, hidden
        31: unknown:18B1           {18B1}   internal, hidden
        32: unknown:1850           {1850}   internal, hidden
        33: unknown:1F03           {1F03}   internal, hidden
        34: unknown:18C0           {18C0}   internal, hidden
        35: LOWRES WHEEL           {2130}   
            Wheel Reports: HID
     Has 9 reprogrammable keys:
         0: LEFT CLICK                , default: LeftClick                   => LEFT CLICK                
             mse, reprogrammable, pos:0, group:1, gmask:1
         1: RIGHT CLICK               , default: RightClick                  => RIGHT CLICK               
             mse, reprogrammable, pos:0, group:1, gmask:1
         2: MIDDLE BUTTON             , default: MiddleMouseButton           => MIDDLE BUTTON             
             mse, reprogrammable, divertable, persistently divertable, pos:0, group:2, gmask:3
         3: BACK AS BUTTON 4          , default: BackEx                      => BACK AS BUTTON 4          
             mse, reprogrammable, divertable, persistently divertable, pos:0, group:2, gmask:3
         4: FORWARD AS BUTTON 5       , default: BrowserForwardEx            => FORWARD AS BUTTON 5       
             mse, reprogrammable, divertable, persistently divertable, pos:0, group:2, gmask:3
         5: unknown:00ED              , default: unknown:00C4                => unknown:00ED              
             mse, reprogrammable, divertable, persistently divertable, pos:0, group:2, gmask:3
         6: LEFT SCROLL AS AC PAN     , default: HorzScrollLeftSet           => LEFT SCROLL AS AC PAN     
             mse, reprogrammable, divertable, persistently divertable, pos:0, group:2, gmask:3
         7: RIGHT SCROLL AS AC PAN    , default: HorzScrollRightSet          => RIGHT SCROLL AS AC PAN    
             mse, reprogrammable, divertable, persistently divertable, pos:0, group:2, gmask:3
         8: unknown:00D7              , default: unknown:00B4                => unknown:00D7              
             divertable, virtual, pos:0, group:3, gmask:0
     Battery: 90%, discharging.
```

**Describe the bug**
While discharging over time, the only levels that solaar show ever shows is 90%, 50%, and 20%.  I have a strong feeling that might be the only levels that are reported by the device and this isn't a solaar issue but is there any way I can tell?  Some additional debug in solaar show?

Anything I can do here to help would be good, I always charge at 50% because I feel 20% is too low, so more granularity would be good, but it might not be possible.

I have my own logs from querying the device every 5 minutes, if that's useful at all.  I've had this device a few years, it's always been like this.  Of course I've removed a ton of duplicate lines from here, but this is essentially the charging status from full to 20% over time with the duplicate levels removed.

Sun Nov 19 10:00:01 EST 2017,1511103603,90
Sat Dec 23 16:00:01 EST 2017,1514062803,50
Sat Dec 23 19:00:01 EST 2017,1514073603,20

These logs are old, but this is the last time I really let it get to 20%, like I said above, I charge at 50%, my https://gitlab.com/khaytsus/mxergo scripts alert me.

For giggles, since I have git master as of right now pulled down, I checked that, same, 90%.  I can keep comparing if more data is useful.

**To Reproduce**
1. Watch solaar show as the device discharges over time
2. Note that it only shows 90, 50, or 20 levels

**Comment by doctor64** [comment]
@khaytsus it will be good if you provide debug logs, but yes, many devices report only limited set of battery states. For example, part of log from M325
```
01:41:17,556    DEBUG [MainThread] logitech_receiver.base: (3) <= w[10 01 040F 000000]
01:41:17,650    DEBUG [MainThread] logitech_receiver.base: (3) => r[11 01 040F 5A050000000000000000000000000000]
01:41:17,650    DEBUG [MainThread] logitech_receiver.hidpp20: device 1 battery 90% charged, next level 5% charge, status 0 = discharging
```
Packet payload decodes as byte0 - BatteryDischargeLevel, byte1 - BatteryDischargeNextLevel, byte2 - Battery status. So 0x5A0500 decodes as current battery - 90%, next reported level - 5%, state - discharging. So, M325 report only 2 battery levels: 90 and 5.

**Comment by pfps** [comment]
`solaar show` just reports what it sees from the device.  Different devices report different battery levels.  There is some information that solaar does not show - many devices also report the next level i.e, at 90% the next level is 50% for your device.  I'm not sure how good these levels are - different batteries have different discharge characteristics and these batteries are not the high-quality, high-price batteries that are found in (decent) laptops.  So I would say that this is not a bug, and that Solaar can't do much better than it currently does.

**Comment by doctor64** [comment]
@pfps Looking into description of feature 0x1000 Battery Status  i note what accordingly to doc device can report `The number of levels the device is capable of reporting to the SW. Min : 2 - Max :100.`. May be it can be added to output of `solaar show` to minimize confusion from users?

**Comment by pfps** [comment]
@doctor64 That's possible.  Not high on my list of things to do, though.  Maybe this issue should be edited and turned into an enhancement request.

**Comment by pfps** [comment]
Fixed by part of PR #802
